### DISCUSSION TITLE
extension can't find lots of modules

### DISCUSSION THREAD

**Discussion by RossBoylan** [microsoft/vscode-discussions#941] [Extension Development QnA] [answered]
I am new to `VSCode` extension development, and generally could use more explanation of the typical development cycle than  the documentation, e.g., https://code.visualstudio.com/api/get-started/your-first-extension and material after, seems to provide.

# Problem

I had installed the `literate` [extension](https://github.com/jesterKing/literate) and decide to try a small tweak.  I cloned the repository and made the tweak.  After attempting unsuccessfully to import from its top level directory using `Developer: Install Extension from Location...` I followed the example and hit `F5` from `extension.ts`.  The initial failures (`tsc` not a recognized command) led me to `npm install -g typescript`.  Then `F5` generated many errors about not being able to find modules, including both basic `node` modules like `fs` and more `VSCode`-specific ones like `vscode` and `markdown-it`.

A [post](https://bobbyhadz.com/blog/typescript-could-not-find-a-declaration-file-for-module#fix---could-not-find-declaration-file-for-module-x-error), mentioned in another thread, provided some tips and I went one round with `npm i --save-dev @types/node`.  This did reduce the number of errors, but there are still 52, and most of the error messages in the transcript don't have a suggestion to install a types module.

# Lament

Considering I already have the extension running, I hoped this would be easier.

# Possible Causes and Workarounds

One possible culprit is that I have an installation of `node.js` that is both outside of `VSCode` and a different version (the internal one seems to be 18 but I installed 20).

I gather from the discussion of publishing that I might be able to make the whole extension into a package and then load it.  But that seems like a cumbersome process just to test some code tweaks.  And I don't know if I can debug the resulting installation.

# P.S.: There's More

This package actually has an additional level of indirection to get to the final code.  The ground truth source code is in the `.literate` files, which must be processed by the extension's `Literate: Process` command to produce the `.ts` files (and maybe some others).  So I put my tweak in the proper `.literate` file and manually issued `Literate: Process` to rebuild the rest.  Those results must then be processed by `tsc`.

**Comment by alefragnani** [comment]
Hi @RossBoylan ,

Building your first extension should be fairly easy, and following the documentation provided in https://code.visualstudio.com/api/get-started/your-first-extension, you will have a Hello World extension running without issues. However, extension authors can be very imaginative on how to bundle their extensions and the dependencies that it requires. This leads to, sometimes complex, development process. To make contributors life easier, extension authors could include a `CONTRIBUTING.md` guide, detailing the _pre-requisites_, _dependencies_, and _instructions on how to build and debug_ the extension. This guide is from one of my extensions https://github.com/alefragnani/vscode-separators/blob/master/CONTRIBUTING.md.

Looking at the source of the **literate** extension you commented, I could noticed that it uses `esbuild` as a bundler. This forces you to install it on your environment to build the extension yourself. VS Code itself provides a documentation that details how you can bundle an extension using `esbuild` (https://code.visualstudio.com/api/working-with-extensions/bundling-extension#using-esbuild), and a quick look at his `scripts` section, it seems the extension followed those instructions. 

Once you are able to build the extension with no errors, hitting <kbd>F5</kbd> should allows you to debug the extension.

About your **lament comment**: _Considering I already have the extension running, I hoped this would be easier._, it is understandable your expectation, but this is _not necessarily_ how installed extension works. In this particular case, since the author used a _bundler_ while packaging the extension, you won't be able to easily change its source, specially because it was minified. So, by default, consider installed and development/debug as two separate things.

I could build and debug the extension, following the following steps:

1. Opened its repo in Codespaces
2. Run `npm i` to install all dependencies
3. Run `npm i --save-dev esbuild` to install `esbuild`
4. Run `npm run compile` just to transpile TS to JS
5. Run `npm run esbuild` to bundle the extension
6. After that, hit `F5`
7. The Extension Development Host opened and I could debug (at least the activation event) while opening a `.literate` file 

I would suggest you to try following these steps, installing the dependencies. If you don't have success, get in touch with the extension author, asking for more detailed information on how build and debug its extension. 

Hope this helps

**Reply by RossBoylan** [reply]
Thank you very much.  Unfortunately, I still seem to get hung up at step 6.  There is a message `Building...` in the middle of the bottom bar with a circle that just keeps spinning.  (I deviated at Step 1, using my local clone of the extension instead of Codespaces.)  When I click on it I see
```
 *  Executing task: npm run watch 


> literate@0.10.2-rb1 watch
> tsc -watch -p ./
# many blank lines

[11:10:17 PM] Starting compilation in watch mode...

# more whitespace, but nothing else
```
I don't know why it's compiling something I just compiled, or why it is in "watch mode", whatever that is.

Perhaps the problem is that I already have the extension loaded?

I'm also not sure which file I should have selected when hitting `F5`.  I picked `extension.ts`, but I guess it's really running `out\main.js`.  And then what file should I set breakpoints in?

So, on the more general issue of my lament, is each extension running its own isolated environment, so the  `literate` extension I installed in the usual way is the only thing that has access to its modules?  It seems as if that would produce a lot of duplication of modules across extensions that use them.  I think your previous answer implies this is only true for extensions distributed as bundles, though, since the docs indicated this is the preferred approach, that must be a lot of them.

**Comment by theReynald** [comment]
Hi @RossBoylan ,

Are you still working on this issue?

**Reply by RossBoylan** [reply]
I've mostly resolved it.  I contacted the extension author and got more [instructions](https://github.com/jesterKing/literate/issues/26).  However, following them at first ran into the same holdup I had encountered [above](https://github.com/microsoft/vscode-discussions/discussions/941#discussioncomment-7665080).  The problem was that the watch wasn't working,  i.e. the build seemed to hang at the launch of the process with the watch, as if `VSCode` was waiting for the process with the watch to exit.  The build task had `"isBackground": true`,

The extension author removed the watch from the build step, and then I was able to build and test.
This is probably not ideal, since I assume the original watch behavior was desired (not the hanging up, but the rebuilding if the files changed).  If you want to see the configuration that hung (for me on Windows) try [v0.11.1](https://github.com/jesterKing/literate/commit/38c485c0d5690eadead8ba8eae0a31cfa4b37cfa)

I noticed there have been a number of problems over the years with getting background tasks to work properly, particularly on MS-Windows, which I was using.  Should that be possible?  Are there any tricks?  In particular there weren't all the `problemMatcher` definitions that the documentation mentioned.

P.S. Extension author took the suggestion and added a `Contributing.md`.

**Reply by alefragnani** [reply, marked as answer]
Hi @RossBoylan ,

Personally, I've never faced issues with the _watch_ while developing VS Code extensions, until I tried to use `esbuild`. In the end, I couldn't make it work, and I decided to rollback the changes, returning to `webpack`. My extensions are pretty small, and didn't make much difference which bundler to use.

Hope this helps
